### PR TITLE
Provide a subdomain with a reversed certificate chain

### LINKED ISSUE #405: Not able to make serve
I just cloned master and tried
```
make serve
```
The docker image could not be built because of an old ruby version:
```
...
Successfully installed ffi-1.11.1
Building native extensions.  This could take a while...
ERROR:  Error installing jekyll:
	jekyll-sass-converter requires Ruby version >= 2.4.0.
Successfully installed sassc-2.2.0
The command '/bin/sh -c gem install jekyll' returned a non-zero code: 1
Makefile:98: recipe for target 'docker-build' failed
make: *** [docker-build] Error 1
```

I tried updating the Dockerfile to start from `ubuntu:18.04` instead of `ubuntu:16.04` (in order to get `ruby2.5` instead of `ruby2.3`) but the command failed again. Probably nginx is also newer in this version of ubuntu, because it rejected a weak certificate:
```
...
Successfully built 9f29b383e9a9
Successfully tagged badssl:latest
docker run -it -p 80:80 -p 443:443 -p 1000-1024:1000-1024 badssl
nginx: [emerg] SSL_CTX_use_certificate("/var/www/badssl/_site/certs/sets/current/gen/chain/wildcard-md5.pem") failed (SSL: error:140AB18E:SSL routines:SSL_CTX_use_certificate:ca md too weak)
Makefile:102: recipe for target 'docker-run' failed
make: *** [docker-run] Error 1
```

### PR DESCRIPTION
**Pull request by g-andrade**
### Problem

Disorderly certificate chains are [a common mistake](https://www.google.com/search?q=site%3Astackoverflow.com+out-of-order+certificate+chain+&oq=site%3Astackoverflow.com+out-of-order+certificate+chain) when setting up web servers and other kinds of TLS-protected services.

Common browsers are clever enough to walk such certificate chains.

However, a strict TLS implementation like [Erlang/OTP's](https://erlang.org/doc/man/ssl.html) does not allow for certificate chains that are not hierarchically ordered without consumer-provided implementations to achieve it.

Providing a subdomain with a disorderly certificate chain would make testing such code easier.

### Solution

This PR proposes the addition of a new subdomain, ~`disorderly-chain.badssl.com`~ `reversed-chain.badssl.com`, whose certificate chain is declared in reverse order, with the intermediate CA placed before the subdomain certificate.

This ensures that any TLS implementations testing their behaviour against the new subdomain need to know how to walk a ~disorderly~ reversed certificate chain.

Resolves #433 ; Closes #405 .

### CODE REVIEW

**Comment by g-andrade** [comment]
I cherry-picked @omahabrad 's fix from PR #433 in order to make docker builds work, hence the CLA warning.

@omahabrad : do you consent to it? If you do, please leave a comment with "@googlebot I consent."

**Comment by omahabrad** [comment]
@googlebot I consent.

**Comment by g-andrade** [comment]
> This is fairly nitpicky, but might I recommend misordered-chain instead of disorderlychanin? The former is more specific, and might be less confusing to people who aren't as familiar with adjectives (rather than adverbs) that end in "-ly".

Fair enough :-) I was undecided between "disordered", "unordered" and "disorderly" - I missed "misordered" entirely.

> In the case of a chain containing 2 certs, it would also be accurate to call it reversed-chain. If you want to test that a client can handle chains that don't go strictly in ancestor order or child order, an additional intermediate would be useful. I don't know if it's easy to get that from a real CA.

You're right; it's not much of a generic case with only two certificates up the chain, just a simple reversal. I'll look into more complex scenarios.

**Comment by christhompson** [comment]
+1 to "misordered-chain" wording for the multiple-intermediate case. Worst case we could make it test-only if we don't think we can get a publicly trusted certificate easily for this.

`reversed-chain` could use a normal certificate, but I seem to recall that nginx will complain if the leaf and intermediate are in the wrong order (key mismatch error). I haven't pulled this branch locally yet to test -- does this start up as expected?

**Comment by christhompson** [comment]
Also, if we add this (test-only or with a real cert if we can get one) I think it would resolve Issue #374

**Comment by g-andrade** [comment]
> This is fairly nitpicky, but might I recommend misordered-chain instead of disorderly

Fixed on ab30874.

**Comment by g-andrade** [comment]
> I haven't pulled this branch locally yet to test -- does this start up as expected?

Sorry for the unreasonably later answer - it does! I'm able to run `make serve` and load `https://badssl.test` locally, on the browser (as well as `https://misorderly-chain.badssl.test`.)

**Comment by g-andrade** [comment]
> In the case of a chain containing 2 certs, it would also be accurate to call it reversed-chain. If you want to test that a client can handle chains that don't go strictly in ancestor order or child order, an additional intermediate would be useful. I don't know if it's easy to get that from a real CA.

A fair point; with that in mind, I've once more renamed the new subdomain - this time to `reversed-chain`.

Would it be good to merge as is? I'm very sorry for not having picked this PR up for quite some time - other things took priority in the mean time.

**Review by lgarron — APPROVED** [review]
I'd be glad to see this added.

Merging and deploying is up to @christhompson , though!

File: Dockerfile [hunk 1]
```diff
@@ -2,14 +2,18 @@
 FROM ubuntu:16.04
 MAINTAINER April King <april@pokeinthe.io>
 EXPOSE 80 443
+RUN apt-get update && apt-get install -y apt-transport-https
+RUN apt-get install -y software-properties-common
+RUN apt-add-repository ppa:brightbox/ruby-ng
```
Comment (lgarron) [inline Dockerfile:7]: Our of curiosity, would you mind catching me up on what each of these are needed for? :-D

Comment (g-andrade) [reply Dockerfile:7]: I cherry-picked these changes from PR #433, which I needed in order to run `make serve`.  To my shame, I didn't think through the details of each line; perhaps the fix can be made simpler?

I can speculate on what some of these changes likely achieve, though:
1. Installing `apt-transport-https` is needed to retrieve packages from PPA repos (guessing here)
2. _I have no clue_ what `software-properties-common` is about
3. `ppa:brightbox/ruby-ng` is a PPA which provides more up to date Ruby versions (also guessing here)

**Comment by g-andrade** [comment]
Supersedes #433 .

**Comment by g-andrade** [comment]
Closes #433 .

**Comment by paulo-ferraz-oliveira** [comment]
Hi.

We could benefit from this change. Any chance it gets merged soon?

Thanks.

**Comment by paulo-ferraz-oliveira** [comment]
Bump.

**Comment by lgarron** [comment]
@christhompson I could merge, but I can't deploy. Do you think it would be worth setting up automatic deploys? We could set up SSH or pull-based deploys using GitHub Actions fairly easily now.

**Comment by christhompson** [comment]
Sorry for the delay on this one. I can merge and get this deployed.

I also filed #462 to track automating deployment, since it's something I've been thinking about occasionally for a while now.

**Comment by christhompson** [comment]
Double-checking this today, it looks like this maybe isn't testing what we intended: the subdomain config is serving the reversed cert chain but currently using the BadSSL intermediate in the chain, and using the private key of the intermediate instead of a leaf cert. This means that it will trigger a `NET::ERR_CERT_AUTHORITY_INVALID` error in Chrome (and similar in Safari/Firefox). 

If we fix this to build a chain using the main leaf cert (`wildcard-main.crt`) and the publicly-trusted intermediate, and then serve using the `leaf-main.key` private key, then nginx (well, OpenSSL under the hood) will refuse to start due to the cert/key mismatch:

> nginx: [emerg] SSL_CTX_use_PrivateKey_file("/etc/keys/leaf-main.key") failed (SSL: error:0B080074:x509 certificate routines:X509_check_private_key:key values mismatch)

Which is good for nginx/openssl users in general, but bad for us trying to intentionally hold it wrong and serve a reversed cert chain. I did some spelunking in nginx and OpenSSL docs, but it doesn't look like there's an easy way to have it load the chain correctly but then serve it incorrectly.

**Comment by lgarron** [comment]
> Which is good for nginx/openssl users in general, but bad for us trying to intentionally hold it wrong and serve a reversed cert chain. I did some spelunking in nginx and OpenSSL docs, but it doesn't look like there's an easy way to have it load the chain correctly but then serve it incorrectly.

Can we hand off a port to another process?

**Comment by christhompson** [comment]
Yeah, we could probably set up something else to handle this case  -- did you have an alternative in mind?

**Comment by christhompson** [comment]
For now I've removed this from the dashboard and homepage to avoid confusion (#470).

**Comment by g-andrade** [comment]
Oh well, it is what it is...

But why does it work locally and not when deployed, since the Docker container is also running nginx?

I've just re-checked that it works locally.

Firstly I get the warning about an unknown issuer (as expected):

Screenshot at 2021-04-03 00-37-47

And once I bypass that, the page is correctly loaded:

Screenshot at 2021-04-03 00-38-01

(But it wasn't working for me in production, either.)

Is there some fundamental difference I'm missing between the production and local (self-signed) chains? (Aside from the issuer being known in the former.)

**Pull request merged**